redis/src/caching/cache_manager.rs

```diff
@@ -0,0 +1,226 @@
+use crate::caching::statistics::Statistics;
+use crate::caching::{CacheConfig, CacheStatistics};
+use crate::cmd::{CommandCacheInformation, CommandCacheInformationByRef};
+use crate::Value;
+use lru::LruCache;
+use std::cmp::min;
+use std::collections::hash_map::DefaultHasher;
+use std::hash::Hasher;
+use std::num::NonZeroUsize;
+use std::ops::Add;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+use tokio::sync::watch::{channel, Receiver, Sender};
+
+type RedisCmd = Vec<u8>;
+type RedisKey = Vec<u8>;
+type LRUCacheShard = LruCache<RedisKey, CacheItem>;
+
+struct CacheCmdEntry {
+    cmd: RedisCmd,
+    value: Option<Value>,
+    receiver: Receiver<Value>,
+}
+
+struct CacheItem {
+    ttl: Instant,
+    value_list: Vec<CacheCmdEntry>,
+}
+
+struct ShardedLRU {
+    shards: Vec<std::sync::Mutex<LRUCacheShard>>,
+}
+impl ShardedLRU {
+    const MAX_SHARD_SIZE: usize = 32;
+    fn new(total_key_size: NonZeroUsize) -> Self {
+        // If total cache size is smaller than max shard size then it won't use sharding.
+        let (shard_size, per_lru_size) = if total_key_size.get() >= Self::MAX_SHARD_SIZE {
+            (
+                Self::MAX_SHARD_SIZE,
+                total_key_size.get() / Self::MAX_SHARD_SIZE,
+            )
+        } else {
+            (1, total_key_size.get())
+        };
+
+        let mut shards = Vec::with_capacity(shard_size);
+        for _ in 0..shard_size {
+            let shard = LruCache::new(NonZeroUsize::new(per_lru_size).unwrap());
+            shards.push(std::sync::Mutex::new(shard));
+        }
+        ShardedLRU { shards }
+    }
+    fn get_shard(&self, key: &[u8]) -> &std::sync::Mutex<LRUCacheShard> {
+        let mut s = DefaultHasher::new();
+        s.write(key);
+        &self.shards[s.finish() as usize % self.shards.len()]
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct CacheManager {
+    lru: Arc<ShardedLRU>,
+    statistics: Arc<Statistics>,
+    pub(crate) cache_config: CacheConfig,
+}
+
+impl CacheManager {
+    pub fn new(cache_config: CacheConfig) -> Self {
+        CacheManager {
+            statistics: Default::default(),
+            lru: Arc::new(ShardedLRU::new(cache_config.cache_size)),
+            cache_config,
+        }
+    }
+
+    pub(crate) async fn get_with_guard_new<'a>(
+        &self,
+        cache_information: &CommandCacheInformationByRef<'a>,
+    ) -> Result<Option<Value>, Sender<Value>> {
+        let mut receiver = {
+            let mut lru_cache_better = self
```

Comment:
what do you mean by better?

Reply:
oops, will be removed (it's coming from an old patch)